storm-core/src/jvm/org/apache/storm/scheduler/blacklist/CircularBuffer.java

```diff
@@ -0,0 +1,174 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.storm.scheduler.blacklist;
+
+import java.io.Serializable;
+import java.util.AbstractCollection;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.ConcurrentModificationException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public final class CircularBuffer<T extends Serializable> extends AbstractCollection<T> implements Serializable {
```

Comment:
I am a little confused why we need all of this.  It feels like we could do all of this with an ArrayBlockingQueue.

```
public final class CircularBuffer<T extends Serializable> extends ArrayBlockingQueue<T> {
    @Override
    public boolean add(T obj) {
        while (!offer(obj)) {
            poll();
        }
    }

    public List<T> toList() {
        return new ArrayList<>(this);
    }
}
```